tools/odbc/connect/connect.cpp

```diff
@@ -0,0 +1,153 @@
+#include "connect.hpp"
+#include "duckdb/main/db_instance_cache.hpp"
+#include <iostream>
+#include <utility>
+
+#if WIN32
+#include <windows.h>
+#endif
+
+using namespace duckdb;
+
+//! The database instance cache, used so that multiple connections to the same file point to the same database object
+DBInstanceCache instance_cache;
+
+bool Connect::SetSuccessWithInfo(SQLRETURN ret) {
+	if (!SQL_SUCCEEDED(ret)) {
+		return false;
+	}
+	if (ret == SQL_SUCCESS_WITH_INFO) {
+		success_with_info = true;
+	}
+	return true;
+}
+
+SQLRETURN Connect::FindMatchingKey(const std::string &input, string &key) {
+	if (seen_config_options.find(input) != seen_config_options.end()) {
+		key = input;
+		return SQL_SUCCESS;
+	}
+
+	auto config_names = DBConfig::GetOptionNames();
+
+	// If the input doesn't match a keyname, find a similar keyname
+	auto msg = StringUtil::CandidatesErrorMessage(config_names, input, "Did you mean: ");
+	return SetDiagnosticRecord(dbc, SQL_SUCCESS_WITH_INFO, "SQLDriverConnect",
+	                           "Invalid keyword: '" + input + "'. " + msg, SQLStateType::ST_01S09, "");
+}
+
+SQLRETURN Connect::FindKeyValPair(const std::string &row) {
+	string key;
+
+	size_t val_pos = row.find(KEY_VAL_DEL);
+	if (val_pos == std::string::npos) {
+		// an equal '=' char must be present (syntax error)
+		return (SetDiagnosticRecord(dbc, SQL_ERROR, "SQLDriverConnect", "Invalid connection string",
+		                            SQLStateType::ST_HY000, ""));
+	}
+
+	SQLRETURN ret = FindMatchingKey(StringUtil::Lower(row.substr(0, val_pos)), key);
+	if (ret != SQL_SUCCESS) {
+		return ret;
+	}
+
+	config_map[key] = row.substr(val_pos + 1);
+	seen_config_options[key] = true;
+	return SQL_SUCCESS;
+}
+
+SQLRETURN Connect::ParseInputStr() {
+	size_t row_pos;
+	std::string row;
+
+	if (input_str.empty()) {
+		return SQL_SUCCESS;
+	}
+
+	while ((row_pos = input_str.find(ROW_DEL)) != std::string::npos) {
```

Comment:
nit: I think this code could be simplified somewhat by using `StringUtil::Split`